### DISCUSSION TITLE
How can I enable Content-Encoding:gzip to Quarkus HTTP

### DISCUSSION THREAD

**Discussion by SetoKaiba** [quarkusio/quarkus#23162] [Q&A] [answered]
I have a file called WebGL.framework.js.gz in resources/META-INF/Build. I want the http serve it with Content-Encoding: gzip. How?

> WebGL.loader.js:1 Unable to parse Build/WebGL.framework.js.gz! This can happen if build compression was enabled but web server hosting the content was misconfigured to not serve the file with HTTP Response Header "Content-Encoding: gzip" present. Check browser Console and Devtools Network tab to debug.

**Comment by mattwelke** [comment]
Isn't this a repeat of this Stack Overflow question?
https://stackoverflow.com/questions/70843270/how-can-i-enable-content-encodinggzip-to-quarkus-http

**Reply by SetoKaiba** [reply]
I thought that Stack Overflow and Discussions are two different community. So I post it there and here.
Is there an answer to my question?

**Comment by Postremus** [comment, marked as answer]
Please try "quarkus.http.enable-compression=true".

https://quarkus.io/guides/rest-json#gzip-support

**Reply by SetoKaiba** [reply]
![image](https://user-images.githubusercontent.com/159540/150934538-2d013799-d2cc-4c64-b28f-b14a1e58884d.png)
This one is the correct one. It's compressed as 87.2kB and it's decompressed as 476kB.
![image](https://user-images.githubusercontent.com/159540/150934786-167f9e6c-a6fe-42e2-b2a6-17abe7bdd259.png)
This one is the wrong one in Quarkus. It's compressed as 87.3kB and it's decompressed as 86.9kB.

It did add Content-Encoding: gzip to it. But the file WebGL.framework.js.gz is already gziped. It seems that the quarkus.http.enable-compression=true make it compress again. I think I just need  to add Content-Encoding to the file with .gz extension.

**Reply by Postremus** [reply]
okay, in that case, a `ContainerResponseFilter` can help, when you use resteasy-reactive. 
For resteasy classic (the quarkus-resteasy extensions), I am not sure if this gets invoked, since files are served from a vert.x handler before the resteasy classic handler has any chance to run, afaik.

```
@Provider
@PreMatching
public class GZIPResponseFilter implements ContainerResponseFilter {
    @Override
    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) throws IOException {
        if (requestContext.getUriInfo().getPath().contains("WebGL.framework.js.gz")) {
            responseContext.getHeaders().putSingle("Content-Encoding", "gzip");
        }
    }
}
```

**Reply by Postremus** [reply]
Alternativly, you might also be able to configure response headers for a specific path directly:
https://quarkus.io/guides/http-reference#additional-http-headers

**Reply by SetoKaiba** [reply]
Hi, I tried the additional http headers way. I find that I can add Content-Encoding to it by additional http headers with a full path.

> quarkus.http.header."Content-Encoding".path=/WebGL/Build/WebGL.framework.js.gz

This is working.

> quarkus.http.header."Content-Encoding".path=/WebGL/Build/*.gz

This is not working.

> quarkus.http.header."Content-Encoding".path=/WebGL/Build/WebGL.data.gz,/WebGL/Build/WebGL.framework.js.gz,/WebGL/Build/WebGL.wasm.gz

This is not working. I have no way to add multiple path.

**Reply by SetoKaiba** [reply]
> okay, in that case, a `ContainerResponseFilter` can help, when you use resteasy-reactive. For resteasy classic (the quarkus-resteasy extensions), I am not sure if this gets invoked, since files are served from a vert.x handler before the resteasy classic handler has any chance to run, afaik.
> 
> ```
> @Provider
> @PreMatching
> public class GZIPResponseFilter implements ContainerResponseFilter {
>     @Override
>     public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) throws IOException {
>         if (requestContext.getUriInfo().getPath().contains("WebGL.framework.js.gz")) {
>             responseContext.getHeaders().putSingle("Content-Encoding", "gzip");
>         }
>     }
> }
> ```

What do you mean by use resteasy-reactive? I change the extension resteasy classic to resteasy-reactive. It seems that my files are still served by vert.x handler. The filter is not triggerd

**Reply by SetoKaiba** [reply]
@Postremus Can you please help me? I found no way to add header to multiple path web requests.

**Reply by Postremus** [reply]
Finally found time to dig deeper into this.
`quarkus.http.header."Header".path` appears to only support one path at a time, no regex can be used.
https://github.com/quarkusio/quarkus/blob/6a3c9234551905d3c8b46447614a9474276c7cee/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L373

I figured out another way though which works with static files. This registers a so filter for the Vert.X request. In the filter itself the response can be changed, i.e. to include a specific custom header. Works with resteasy classic and resteasy-reactive.
```
import io.quarkus.vertx.http.runtime.filters.Filters;
import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;
@ApplicationScoped
public class FilterRegistrator {
void setupFilter(@Observes Filters filters) {
        filters.register((rc) -> {
            rc.next();

            if (rc.normalizedPath().contains("/index.html") || rc.normalizedPath().contains("/logo.png") || rc.normalizedPath().contains("/favicon.ico")) {
                rc.response().headers().add("CUSTOM-HEADER", "VALUE");
            }

        }, 0);
    }
```

**Reply by SetoKaiba** [reply]
@Postremus Thank you. It does help.

**Reply by mattwelke** [reply]
@SetoKaiba You should post the answer in https://stackoverflow.com/questions/70843270/how-can-i-enable-content-encodinggzip-to-quarkus-http too so that people who find that SO don't assume the problem isn't solvable, not realizing it was solved here.

**Reply by SetoKaiba** [reply]
@Postremus It seems that the filter is not reliable. Did I miss something? Could you please check this link. If I ctrl + f5 to refresh it again and again without cache. The content-encoding: br is missing sometimes. I don't know whether it's a bug or not. The output is always printed. But the content-encoding is missing sometimes.
https://live.kaiba.net:8443/WebGL2/Build/WebGL2.framework.js.br

```
package net.kaiba;


import io.quarkus.vertx.http.runtime.filters.Filters;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;

@ApplicationScoped
public class FilterRegistrator {

    void setupFilter(@Observes Filters filters) {
        filters.register((rc) -> {
            rc.next();

            if (rc.normalizedPath().contains(".gz")) {
                System.out.println(rc.normalizedPath() + ":gz");
                rc.response().headers().add("content-encoding", "gzip");
            }

            if (rc.normalizedPath().contains(".br")) {
                System.out.println(rc.normalizedPath() + ":br");
                rc.response().headers().add("content-encoding", "br");
            }

            if (rc.normalizedPath().contains(".wasm")) {
                System.out.println(rc.normalizedPath() + ":wasm");
                rc.response().headers().add("content-type", "application/wasm");
            }

        }, 0);
    }
}
```

I tried undertow-handlers. But it can't override the content-type.
https://quarkus.io/guides/http-reference#undertow-handlers-conf

I made a simple enough project with only FilterRegistrator and the https support. I can always reproduce with just ctrl + F5 to refresh the page.

**Reply by SetoKaiba** [reply]
The rc.next() should be the last operation. The missing sometimes problem is fixed.
https://github.com/quarkusio/quarkus/issues/24978
